src/Compilers/CSharp/Portable/Binder/Semantics/Conversions/ConversionsBase.cs

```diff
@@ -1578,6 +1583,99 @@ private static bool HasAnonymousFunctionConversion(BoundExpression source, TypeS
+            }
+            else if (isSpanType(compilation, destination, WellKnownType.System_ReadOnlySpan_T, out elementType))
+            {
+                return ConstructibleCollectionTypeKind.ReadOnlySpan;
+            }
+            else if (implementsIEnumerable(compilation, destination))
+            {
+                elementType = null;
+                return ConstructibleCollectionTypeKind.CollectionInitializer;
+            }
+            else if (isListInterface(compilation, destination, out elementType))
+            {
+                return ConstructibleCollectionTypeKind.ListInterface;
+            }
+
+            elementType = null;
+            return ConstructibleCollectionTypeKind.None;
+
+            static bool isSpanType(CSharpCompilation compilation, TypeSymbol targetType, WellKnownType spanType, [NotNullWhen(true)] out TypeSymbol? elementType)
+            {
+                if (targetType is NamedTypeSymbol { Arity: 1 } namedType
+                    && TypeSymbol.Equals(namedType.OriginalDefinition, compilation.GetWellKnownType(spanType), TypeCompareKind.AllIgnoreOptions))
+                {
+                    elementType = namedType.TypeArgumentsWithAnnotationsNoUseSiteDiagnostics[0].Type;
+                    return true;
+                }
+                elementType = null;
+                return false;
+            }
+
+            static bool implementsIEnumerable(CSharpCompilation compilation, TypeSymbol targetType)
+            {
+                ImmutableArray<NamedTypeSymbol> allInterfaces;
+                switch (targetType.TypeKind)
+                {
+                    case TypeKind.Class:
+                    case TypeKind.Struct:
+                        allInterfaces = targetType.AllInterfacesNoUseSiteDiagnostics;
+                        break;
+                    case TypeKind.TypeParameter:
+                        allInterfaces = ((TypeParameterSymbol)targetType).AllEffectiveInterfacesNoUseSiteDiagnostics;
+                        break;
+                    default:
+                        return false;
+                }
+
+                // This implementation differs from Binder.CollectionInitializerTypeImplementsIEnumerable().
+                // That method checks for an implicit conversion from IEnumerable to the collection type,
+                // but that would allow: Nullable<StructCollection> s = [];
+                // PROTOTYPE: Perhaps adjust the behavior of Binder.CollectionInitializerTypeImplementsIEnumerable()
+                // and use that method instead.
+                var ienumerableType = compilation.GetSpecialType(SpecialType.System_Collections_IEnumerable);
+                return allInterfaces.Any(static (a, b) => a.Equals(b, TypeCompareKind.AllIgnoreOptions), ienumerableType);
+            }
+
+            static bool isListInterface(CSharpCompilation compilation, TypeSymbol targetType, [NotNullWhen(true)] out TypeSymbol? elementType)
```

Comment:
It feels like [spec](https://github.com/dotnet/csharplang/blob/main/proposals/collection-literals.md#interface-translation) is out of sync with implementation here.

> If `T` is some interface `I<T1>` where that interface is implemented by `List<T1>`...

The spec implies that if `T` is `IEnumerable<string>`, and `List<string>` implements `IEnumerable<string>`, then we use this particular translation strategy. But what the implementation is actually checking is, does the original definition `List<T_FromList>` implement `IEnumerable<T_FromList>`. In other words, in the implementation, a rogue definition of `List<T>` which directly implements `IEnumerable<string>` won't use this translation strategy, even though spec says to do so.

I think this implementation is fine, and the spec should be adjusted to match it.